bokehjs/src/coffee/models/annotations/color_bar.coffee

```diff
@@ -0,0 +1,401 @@
+      # image_width = legend_width - legend_padding * 2
+
+    ctx.restore()
+
+    location = @mget('location')
+    h_range = @plot_view.frame.get('h_range')
+    v_range = @plot_view.frame.get('v_range')
+
+    if _.isString(location)
+      switch location
+        when 'top_left'
+          x = h_range.get('start') + legend_margin
+          y = v_range.get('end') - legend_margin
+        when 'top_center'
+          x = (h_range.get('end') + h_range.get('start'))/2 - legend_width/2
+          y = v_range.get('end') - legend_margin
+        when 'top_right'
+          x = h_range.get('end') - legend_margin - legend_width
+          y = v_range.get('end') - legend_margin
+        when 'right_center'
+          x = h_range.get('end') - legend_margin - legend_width
+          y = (v_range.get('end') + v_range.get('start'))/2 + legend_height/2
+        when 'bottom_right'
+          x = h_range.get('end') - legend_margin - legend_width
+          y = v_range.get('start') + legend_margin + legend_height
+        when 'bottom_center'
+          x = (h_range.get('end') + h_range.get('start'))/2 - legend_width/2
+          y = v_range.get('start') + legend_margin + legend_height
+        when 'bottom_left'
+          x = h_range.get('start') + legend_margin
+          y = v_range.get('start') + legend_margin + legend_height
+        when 'left_center'
+          x = h_range.get('start') + legend_margin
+          y = (v_range.get('end') + v_range.get('start'))/2 + legend_height/2
+        when 'center'
+          x = (h_range.get('end') + h_range.get('start'))/2 - legend_width/2
+          y = (v_range.get('end') + v_range.get('start'))/2 + legend_height/2
+    else if _.isArray(location) and location.length == 2
+      [x, y] = location
+
+    sx = @plot_view.canvas.vx_to_sx(x)
+    sy = @plot_view.canvas.vy_to_sy(y)
+
+    image_sx = sx + legend_padding
+    image_sy = sy + legend_padding + title_height
+
+    return {sx: sx, sy: sy, width: legend_width, height: legend_height, image_sx: image_sx, image_sy: image_sy, image_width: image_width, image_height: image_height}
+
+  render: () ->
+    if @model.visible == false
+      return
+
+    ctx = @plot_view.canvas_view.ctx
+    ctx.save()
+
+    if @model.panel?
+      panel_offset = @_get_panel_offset()
+      ctx.translate(panel_offset.x, panel_offset.y)
+
+    @_draw_bbox(ctx)
+    @_draw_image(ctx)
+    @_draw_major_ticks(ctx)
+    @_draw_minor_ticks(ctx)
+    @_draw_major_labels(ctx)
+    @_draw_title(ctx)
+    ctx.restore()
+
+  _draw_bbox: (ctx) ->
+    if not @visuals.background_fill.doit
+      return
+
+    geom = @compute_legend_bbox()
+    @visuals.background_fill.set_value(ctx)
+    ctx.save()
+    ctx.fillRect(geom.sx, geom.sy, geom.width, geom.height)
+    ctx.restore()
+
+  _draw_image: (ctx) ->
+    geom = @compute_legend_bbox()
+    @visuals.border_line.set_value(ctx)
+    ctx.save()
+    ctx.setImageSmoothingEnabled(@mget('smooth_scale'))
```

Comment:
btw. I encourage to use `@model.attr_name` instead of `@mget("attr_name")`. At some point I would like to get rid of by-string references to attributes (e.g. to aid code linting).